### ISSUE TITLE
Bug: Form submit is triggered when a non submit type button is pressed

### ISSUE THREAD

**Issue by sylsaint** [facebook/react#26711]
React version: 18.2.0

## Steps To Reproduce

1. write a form component with two buttons which are "go next" and "submit" alternatively show by a control value
2. press the go next button
3. form submit is triggered
4. page reloads

Link to code example: https://codesandbox.io/s/flamboyant-kirch-1dufrh?file=/src/App.js

## The current behavior

form submit is triggered after pressing "go next" button

## The expected behavior

form submit should not be triggered after pressing "go next" button

## Cause

Demo code

```javascript
import "./styles.css";
import { useCallback, useState } from "react";

export default function App() {
  const [showSubmit, setShowSubmit] = useState(false);
  const handleSubmit = useCallback((evt) => {
    console.log("submit detected: ", evt);
  }, []);
  return (
    <div className="App">
      <h1>Form error demo</h1>
      <form onSubmit={handleSubmit}>
        {showSubmit === false ? (
          <button type="button" onClick={() => setShowSubmit(true)}>
            go next
          </button>
        ) : (
          <button type="submit">submit</button>
        )}
      </form>
    </div>
  );
}
```

1. When "go next" button is pressed, onClick event handler is triggered
2. showSubmit is set to be true, "submit" button shows
3. React reuse the same object for both "go next" and "submit", just modify its properties
4. Because button pressing event has default action with it
5. When event handler receives the event, event target node's reference is not modified, but its properties has been modified
6. So submit is triggered, cause page to reload

In summary, event target is modified during propagation and before submit event handler

### Current workaround

1. set `evt.preventDefault()` in "go next" button handler
2. set different key for "go next" and "submit"
3. delay the `setShowSubmit(true)`, such as `setTimeout(() => { setShowSubmit(true); }, 100);`

### Final solution expected

event target should get a snapshot before propagating, should not be modified during propagating.

**Comment by gaearon** [comment]
Hm yea this seems confusing but matches how the DOM works. I think I'll close this.

**Comment by sylsaint** [comment]
@nmain Your code is based on the the same DOM node, it's ok. But in react they are actually different node, during render process, react reuse last DOM node, just modify its properties.

**Comment by sylsaint** [comment]
@gaearon Yes, it's confusing. But actually it's caused by react that reused the same DOM node for different buttons.

**Comment by nmain** [comment]
@sylsaint React's reuse of the DOM node here is expected given the reconciliation algorithm's specifications and constraints, and the way you've structured your JSX.  You're returning the same VDOM structure each time with a `<button>` in the same place, but only the attributes on it are different.

The docs go into a lot of depth on this sort of scenario https://react.dev/learn/preserving-and-resetting-state
